### PR TITLE
BWA-MEM support

### PR DESCRIPTION
**Pull request by JohnMCMa**
This pull request addresses issue #1756. 

It adds a new class `BwaSampeCommandline` to Bio.Sequencing.Applications._bwa.py as a wrapper to `bwa mem` with usage similar to the other classes in the file. I also added an unittest, `test_mem`, for `BwaSampeCommandline`.

The code has been tested under python 3.6.1 and style-checked using `flake8` and pre-commit hooks.

- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file and understand that AppVeyor and
TravisCI will be used to confirm the Biopython unit tests and ``flake8`` style
checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

### CODE REVIEW

File: Bio/Sequencing/Applications/_bwa.py [hunk 1]
```diff
@@ -380,7 +380,118 @@ def __init__(self, cmd="bwa", **kwargs):
...
+
+    Run a BWA-MEM alignment, with single- or paired-end reads, equivalent to::
+
+        $ bwa mem [...] <in.db.fasta> <in1.fq> <in2.fq> > <out.sam>
+
+    See http://bio-bwa.sourceforge.net/bwa.shtml for details.
+
+    Examples
+    --------
+    >>> from Bio.Sequencing.Applications import BwaMemCommandline
```
Comment (peterjc) [inline Bio/Sequencing/Applications/_bwa.py]: You need a blank line before the doctest for it to be formatted correctly as RST in our API documentation generated from the source code.

**Comment by peterjc** [comment]
It looks like the automated testing got stuck due to the merge conflict. Are you familiar with using ``git rebase origin/master``?

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+-   repo: https://github.com/pre-commit/pre-commit-hooks
```
Comment (peterjc) [inline .pre-commit-config.yaml]: Where did this file come from? There may be a case for adding this but we could discuss that on a separate issue / pull request.

Comment (JohnMCMa) [reply .pre-commit-config.yaml]: This is just to install the hooks. I will remove this file.

File: Bio/Sequencing/Applications/_bwa.py (hunk 1 above)
Comment (JohnMCMa) [reply Bio/Sequencing/Applications/_bwa.py]: All classes in the file lack the blank lines you mentioned; will add them.

**Comment by peterjc** [comment]
It looks like the blank lines before each doctest broke an automated style check - right now the original style is working within epydoc, e.g. http://biopython.org/DIST/docs/api/Bio.Sequencing.Applications._samtools.SamtoolsViewCommandline-class.html looks fine.

Maybe I was mis-remembering a related problem, or is is specific to using Sphinx (which we don't yet do but are looking at using instead)?

Sorry about that confusion.

**Comment by JohnMCMa** [comment]
@peterjc 
I reverted any edits on NEWS.rst to handle any merge conflict issues and moved on to other edits based on your recommendation. I will handle the conflict when I get to a workstation that I can safely use commandline git.

**Comment by peterjc** [comment]
Looks good to me - @saketkc could you take a quick look at this for us please?

**Comment by saketkc** [comment]
Is it possible to keep the help string text to 80 characters as well? The original `_bwa.py` also doesn't enforce this. In hindsight I should have kept it to 80 characters, happy to send a separate PR for that.

Looks good to me otherwise! Thanks @JohnMCMa!

**Comment by peterjc** [comment]
OK, I'll merge this as is - looking at the line lengths can come later.

**Comment by peterjc** [comment]
Thank you both 👍

**Pull request merged**